source/rsa.c

```diff
@@ -0,0 +1,273 @@
+        case AWS_CAL_RSA_ENCRYPTION_OAEP_SHA512:
+            return key_size_in_bytes - 2 * (512 / 8) - 2;
+        default:
+            AWS_FATAL_ASSERT("Unsupported RSA Encryption Algorithm");
+    }
+
+    return 0;
+}
+
+int aws_rsa_key_pair_encrypt(
+    const struct aws_rsa_key_pair *key_pair,
+    enum aws_rsa_encryption_algorithm algorithm,
+    struct aws_byte_cursor plaintext,
+    struct aws_byte_buf *out) {
+    AWS_PRECONDITION(key_pair);
+    AWS_PRECONDITION(out);
+
+    if (AWS_UNLIKELY(aws_rsa_key_pair_max_encrypt_plaintext_size(key_pair, algorithm) < plaintext.len)) {
+        AWS_LOGF_ERROR(AWS_LS_CAL_RSA, "Unexpected buffer size. For RSA, ciphertext  is expected to match block size");
+        return aws_raise_error(AWS_ERROR_CAL_BUFFER_TOO_LARGE_FOR_ALGORITHM);
+    }
+
+    return key_pair->vtable->encrypt(key_pair, algorithm, plaintext, out);
+}
+
+AWS_CAL_API int aws_rsa_key_pair_decrypt(
+    const struct aws_rsa_key_pair *key_pair,
+    enum aws_rsa_encryption_algorithm algorithm,
+    struct aws_byte_cursor ciphertext,
+    struct aws_byte_buf *out) {
+    AWS_PRECONDITION(key_pair);
+    AWS_PRECONDITION(out);
+
+    if (AWS_UNLIKELY(ciphertext.len != (key_pair->key_size_in_bits / 8))) {
+        AWS_LOGF_ERROR(AWS_LS_CAL_RSA, "Unexpected buffer size. For RSA, ciphertext is expected to match block size.");
+        return aws_raise_error(AWS_ERROR_INVALID_ARGUMENT);
+    }
+
+    return key_pair->vtable->decrypt(key_pair, algorithm, ciphertext, out);
+}
+
+int aws_rsa_key_pair_sign_message(
+    const struct aws_rsa_key_pair *key_pair,
+    enum aws_rsa_signature_algorithm algorithm,
+    struct aws_byte_cursor digest,
+    struct aws_byte_buf *out) {
+    AWS_PRECONDITION(key_pair);
+    AWS_PRECONDITION(out);
+
+    return key_pair->vtable->sign(key_pair, algorithm, digest, out);
+}
+
+int aws_rsa_key_pair_verify_signature(
+    const struct aws_rsa_key_pair *key_pair,
+    enum aws_rsa_signature_algorithm algorithm,
+    struct aws_byte_cursor digest,
+    struct aws_byte_cursor signature) {
+    AWS_PRECONDITION(key_pair);
+
+    return key_pair->vtable->verify(key_pair, algorithm, digest, signature);
+}
+
+size_t aws_rsa_key_pair_block_length(const struct aws_rsa_key_pair *key_pair) {
+    AWS_PRECONDITION(key_pair);
+    return key_pair->key_size_in_bits / 8;
+}
+
+size_t aws_rsa_key_pair_signature_length(const struct aws_rsa_key_pair *key_pair) {
+    AWS_PRECONDITION(key_pair);
+    return key_pair->key_size_in_bits / 8;
+}
+
+int aws_rsa_key_pair_get_public_key(
+    const struct aws_rsa_key_pair *key_pair,
+    enum aws_rsa_key_export_format format,
+    struct aws_byte_buf *out) {
+    (void)format; /* ignore format for now, since only pkcs1 is supported. */
+    AWS_PRECONDITION(key_pair);
+    AWS_PRECONDITION(out);
+
+    if (key_pair->pub.len == 0) {
+        return aws_raise_error(AWS_ERROR_PLATFORM_NOT_SUPPORTED);
+    }
+
+    aws_byte_buf_init_copy(out, key_pair->allocator, &key_pair->pub);
```

Comment:
The docs say this "Appends key bytes to out.", which is what encrypt and decrypto do

Do one of:
- don't init the byte-buf, just append(), like  encrypt and decrypt do
OR
- Make it super clear in the docs that byte-buf should be passed in uninitialized